idaes/models/properties/general_helmholtz/helmholtz_parameters.py

```diff
@@ -0,0 +1,454 @@
+            self.model.delta = delta_v
+            P_v = pyo.value(
+                rho_v * self.R * T * (1 + self.model.delta * self.model.phir_d)
+            )
+            print(
+                f"{T:7.3f}, {T - 273.15: 8.3f}, {rho_l:14.4f}, {rho_v:14.4f}, {P_v:9.4f}"
+            )
+        print("=====================================================================\n")
+
+    def calculate_reference_offset(self, delta, tau, s0, h0):
+        """Given delta and tau for a reference state and the reference entropy and enthalpy, calculate
+        the reference state offset parameters. Since temperature and density are independent of reference
+        state, they can be calculated at a new reference state.
+
+        Args:
+            delta (float): reduced density at new reference state
+            tau (float): 1/reduced temperature at new reference state
+            s0 (float): Entropy at new reference state [kJ/kg/K]
+            h0 (float): Enthalpy at new reference state [kJ/kg]
+
+        Returns:
+            (tuple): Offset for given reference state
+        """
+        self.model.tau = tau
+        self.model.delta = delta
+        s1 = pyo.value(
+            self.model.tau * (self.model.phii_t + self.model.phir_t)
+            - self.model.phii
+            - self.model.phir
+        )
+        n1_off = s1 - s0
+        h1 = pyo.value(
+            1.0
+            / self.model.tau
+            * (
+                1
+                + self.model.tau * (self.model.phii_t + self.model.phir_t)
+                + self.model.delta * self.model.phir_d
+            )
+        )
+        n2_off = h0 - h1
+        return (n1_off, n2_off)
+
+    def write_model(self, model, model_name, expressions=None):
+        """Write an NL file and create an expression and variable map for the EoS model or just a
+        variable map for transport property expressions where there is only one expression per file.
+
+        Args:
+            model (ConcreteModel): a Pyomo model
+            model_name (str): the model name used in the NL file
+
+        Returns:
+            tuple: NL file, expression map, variable map for EOS
+        """
+        nl_file, smap_id = model.write(f"{self.comp}_expressions_{model_name}.nl")
+        for v in model.component_data_objects(pyo.Var):
+            v.unfix()
+        smap = model.solutions.symbol_map[smap_id]
+        var_map = [1000] * 4
+        for s, c in smap.bySymbol.items():
+            if s.startswith("v"):
+                j = int(s[1:])
+                var_map[j] = self.variables[c().name]
+        if expressions is not None:
+            expr_map = [0] * len(expressions)
+            for s, c in smap.bySymbol.items():
+                if s.startswith("o"):
+                    i = expressions[c().name]
+                    j = int(s[1:])
+                    expr_map[i] = j
+            return nl_file, expr_map, var_map
+        return nl_file, None, var_map
+
+    def write(self):
```

Comment:
For testing purposes, you might want to consider having this method write to a stream (which by default would be a file). However, for testing you could write to a `StringIO` object and then test that to avoid having to write and then read a file.

Reply:
I added a dry run option that skips writing the files.  I'll come back to this if I have to.  A big chunk of these writes are nl files, and I not sure how best to test now.  I'm testing the expression before writing to nl files though.  Since this is only needed to write files for new components, it's not critical. The generated files will be tested when they are generated. I'm going to see if I can get away with not testing this method for now.